1.18 - fix for lib: prefix handling

kwvanderlinde opened the pull request:
### Identify the Bug or Feature request

Fixes #5706

### Description of the Change

A few places have been updated to use the callable location instead of the plain location from the `MacroLocation`:
- When resolving `@this` in `defineFunction()`
- In the tooltips for `/say` and `/gm`.
- In the `"source"` value returned by `getMacroContext()`

A few other places directly handle `lib:` prefixes but weren't correct, particularly for tokens with strange but valid names like `Lib:Lib:Hero`:
- `LibraryTokenManager#getLibrary()` once again always adds the `lib:` prefix rather than doing so conditionally. All callers were checked to make sure they pass the namespace, thereby removing the need for the check in the first place.
- `MapToolLineParser#runMacro()` does not remove the `lib:` prefix since it is guaranteed to have the namespace already from `MacroLocation#getLocation()`.
- Similarly, functions like `getLibProperty()` do not try to remove the `lib:` prefix from the result of `MacroLocation#getLocation()`, but they do still need to be able to remove it from a user-supplied lib:token name.

Similar to the `Lib:Lib:Hero` naming issue, macros named like `macro/example` on lib:tokens had the same sort of problem when trying to reference them via URIs with the `macro()` roll option, e.g., `[r, macro(lib://Dragon/macro/macro/example): ""]`. Add-on macros, on the other hand, simply could not be referenced via `macro()` since a full URI could not be resolved by add-ons. Both of these have been fixed. As part of this fix, URI `MacroLocation` and library `MacroLocation` need to be resolved differently, so there is a new method `Library#getMTScriptMacroInfoForUriPath(String)` that supports the URI case, while `Library#getMTScriptMacroInfo(String)` continues to support the library case (among other things).

### Possible Drawbacks

Should be none

### Documentation Notes

N/A

### Release Notes

- Fix various places to correctly handle the `lib:` prefix on lib:tokens.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5708)
<!-- Reviewable:end -->

kwvanderlinde committed:
Fix a few more places that need to use callable location

- When resolving `@this` for `defineFunction()`
- When displaying the source macro for `/say` and `/gm` tooltips
- When deciding the `"source"` value for `getMacroContext()`

kwvanderlinde committed:
Fix incorrect checks for `lib:` prefixes on token names

Lib:tokens can be validly named like `lib:lib:Hero`, so the library namespace can still contain `lib:`. Thus, we can not
dynamically check whether the prefix has been removed, but must be explicit about requiring a namespace vs a full
lib:token name.

Fixes are applied to:
- `LibraryTokenManager#getLibrary()`, to always - not conditionally - add the `lib:` prefix when building the full token
  name
- `MapToolLineParser#runMacro()`, to not remove the `lib:` prefix from the macro location as it has already been
  stripped.
- Lib:token properties (`getLibProperty()` et al), to not strip the `lib:` prefix from the macro location (it still has
  to strip it from a user-supplied lib:token name, though).

cwisniew commented on src/main/java/net/rptools/maptool/client/MapToolLineParser.java:
```diff
@@ -1240,7 +1240,7 @@ public String runMacro(
               macro = mloc.getUri().toString();
             }
             case library -> {
-              namespace = mloc.getLocation().replaceFirst("^(?i)lib:", "");
```
Does fetch from a macro and static file work after this change? It may break it

kwvanderlinde replied:
I will give it this a test to be sure.

It _shouldn't_ break since `getLocation()` should already have the `lib:` stripped off.

kwvanderlinde replied:
I've validated that `fetch()` still works with both `lib://` URIs and `macro:` URIs, and that `macro()` still works with `lib://` URIs.

I'll try to get a test for static files, but I don't do a whole lot with add-ons yet so that might take me a bit.

kwvanderlinde replied:
Okay, I've now validated that fetch() works for both lib:tokens and add-on libraries, and checked that it can both call macros and retrieve static files.

For lib:tokens, this kind of call also works to invoke the `doFetch@Lib:Dragon`:
```
[h, macro("lib://Dragon/macro/doFetch"): ""]
```
For add-on libraries, I assume this is the correct syntax:
```
[h, macro("lib://Dragon/doFetch"): ""]
```
But it just errors out with "Unknown macro "doFetch"." I don't know yet if this means something is broken right now, or if I'm just doing something wrong with add-on - the latter comes with a high probability.

kwvanderlinde replied:
Figured it out - the URI case is not correct. Will push a fix for that as well later today.

kwvanderlinde replied:
To expand on the current problem: URIs cannot be unambiguously resolved to a library namespace and macro name without first knowing whether the referenced library is a token or an add-on. This is because the string `lib://Dragon/macro/example` could either reference:
- the `macro/example` macro in the `Dragon` add-on; or
- the `example` macro in the `Lib:Dragon` token.

Moreover `Library#getMTScriptMacroInfo()` also has completely different expectations depending on the implementation:
- Add-ons expect only the macro name proper to be provided, so that `public/` can be prepended to it and be resolved to a macro file name.
- Tokens need the whole URI so that the `macro/` part can be stripped off.

On top of that, the way `LibraryToken#getMTScriptMacroInfo()` parses the URI means macros named like `macro/example` cannot be called using the expected URI `lib://Dragon/macro/macro/example`.

^ All that is to explain the need for my last commit here. With these fixes, all my `lib:` URIs in different scenarios are now working.

cwisniew replied:
> On top of that, the way `LibraryToken#getMTScriptMacroInfo()` parses the URI means macros named like `macro/example` cannot be called using the expected URI `lib://Dragon/macro/macro/example`.
> 
Yeah, because a macro and a property on a lib: token can have the same name it requires the macro/ and property/ "directory" in the URL to properly distinguish :(

kwvanderlinde committed:
Make lib: URIs for macros be handled consistently

The new `Library#getMTScriptMacroInfoForUriPath(String)` explicitly resolves macro paths (which have a leading `macro/`
for lib:tokens) so that `Library#getMTScriptMacroInfo(String)` is free to resolve pure macro names. This liberates
`LibraryToken#getMTScriptMacroInfo` from needing to parse full URIs.

Corresponding changes have been made to the line parser so that library namespace are handled consistently between URI
and library `MacroLocation`s, while permitting the appropriate macro resolution strategies to be used depending on which
kind of `MacroLocation` we are resolving.
